src/main/java/org/sunbird/hubservices/daoimpl/GraphDao.java

```diff
@@ -61,72 +59,85 @@ public Boolean upsertNode(Node node) throws Exception {
                 logger.info("user node with id {} created successfully ", node.getId());
             }
         } catch (Exception e) {
-            transaction.rollbackAsync().toCompletableFuture().get();
             logger.error("user node creation failed : ", e);
             return Boolean.FALSE;
 
-        } finally {
-            try {
-                if (session != null && transaction != null) {
-                    transaction.close();
-                    session.close();
-                }
-            } catch (Exception e) {
-                logger.error("Session Closed : ", e);
-                return Boolean.FALSE;
-            }
         }
-
         return Boolean.TRUE;
     }
 
     @Override
     public Boolean upsertRelation(Node nodeFrom, Node nodeTo, Map<String, String> relationProperties) throws Exception {
-        try (Session session = neo4jDriver.session()) {
-            try (Transaction transaction = session.beginTransaction()) {
-                Map<String, Object> parameters = new HashMap<>();
-                parameters.put("fromUUID", nodeFrom.getId());
-                parameters.put("toUUID", nodeTo.getId());
-                parameters.put(Constants.Graph.PROPS.getValue(), relationProperties);
+        boolean isUpserted = Boolean.FALSE;
+        try (Session session = neo4jDriver.session(); Transaction transaction = session.beginTransaction()) {
+            Map<String, Object> parameters = new HashMap<>();
+            parameters.put("fromUUID", nodeFrom.getId());
+            parameters.put("toUUID", nodeTo.getId());
+            parameters.put(Constants.Graph.PROPS.getValue(), relationProperties);
 
-                String queryNodeExistWithReverseEdge = "MATCH (n:" + label + ")<-[r:connect]-(n1:" +
-                        label + ") WHERE n.id = $fromUUID AND n1.id = $toUUID " + "RETURN n,n1";
-
-                Statement statement = new Statement(queryNodeExistWithReverseEdge, parameters);
-                StatementResult result = transaction.run(statement);
-                int recordSize = result.list().size();
-                result.consume();
-                if (recordSize != 0) {
-                    throw new GraphException(ErrorCode.NODES_WITH_REVERSE_EDGE_EXISTS.name(),
-                            "Nodes exists with reverse edge, new relation cannot be created!");
-                }
+            String queryNodeExistWithReverseEdge = "MATCH (n:" + label + ")<-[r:connect]-(n1:" +
+                    label + ") WHERE n.id = $fromUUID AND n1.id = $toUUID " + "RETURN n,n1";
 
+            Statement statement = new Statement(queryNodeExistWithReverseEdge, parameters);
+            StatementResult result = transaction.run(statement);
+            int recordSize = result.list().size();
+            result.consume();
+            if (recordSize != 0) {
+                if (logger.isDebugEnabled())
+                    logger.debug(nodeFrom.getId(), nodeTo.getId());
```

Comment:
Just add more information--> printing only fromId and toId may not be helpful...